### PR TITLE
feat(tLN): instantiate subscription supervision

### LINKED ISSUE #20: (tLN) (tControl) Instantiate Supervision
We currently have:

```ts
/**
 * Returns an array with a single Insert Edit to create a new
 * supervision element for the given GOOSE/SMV message and subscriber IED.
 *
 * @param controlBlock The GOOSE or SMV message element
 * @param subscriberIED The subscriber IED
 * @returns an empty array if instantiation is not possible or an array with a single Create action
 */
export function instantiateSubscriptionSupervision(
  controlBlock: Element | undefined,
  subscriberIED: Element | undefined
): Insert[] {
  const supervisionType =
    controlBlock?.tagName === 'GSEControl' ? 'LGOS' : 'LSVS';
  if (
    !controlBlock ||
    !subscriberIED ||
    !isSupervisionAllowed(controlBlock, subscriberIED, supervisionType)
  )
    return [];
  const availableLN = findOrCreateAvailableLNInst(
    controlBlock,
    subscriberIED,
    supervisionType
  );
  if (
    !availableLN ||
    !isSupervisionModificationAllowed(subscriberIED, supervisionType)
  )
    return [];

  const edits: Insert[] = [];
  // If creating new LN element
  if (!availableLN.parentElement) {
    const parent = subscriberIED.querySelector(
      `LN[lnClass="${supervisionType}"]`
    )?.parentElement;
    if (parent) {
      // use Insert edit for supervision LN
      edits.push({
        parent,
        node: availableLN,
        reference:
          parent!.querySelector(`LN[lnClass="${supervisionType}"]:last-child`)
            ?.nextElementSibling ?? null,
      });
    }
  }

  // Insert child elements
  const supervisionName = supervisionType === 'LGOS' ? 'GoCBRef' : 'SvCBRef';

  let doiElement = availableLN.querySelector(`DOI[name="${supervisionName}"]`);
  if (!doiElement) {
    doiElement = subscriberIED.ownerDocument.createElementNS(
      SCL_NAMESPACE,
      'DOI'
    );
    doiElement.setAttribute('name', supervisionName);
    edits.push({
      parent: availableLN!,
      reference: null,
      node: doiElement,
    });
  }

  let daiElement = doiElement.querySelector(`DAI[name="setSrcRef"]`);
  if (!daiElement) {
    daiElement = subscriberIED.ownerDocument.createElementNS(
      SCL_NAMESPACE,
      'DAI'
    );
    const srcValRef = subscriberIED.querySelector(
      `LN[lnClass="${supervisionType}"]>DOI[name="${supervisionName}"]>DAI[name="setSrcRef"]`
    );
    daiElement.setAttribute('name', 'setSrcRef');

    // transfer valKind and valImport from first supervision instance if present
    if (srcValRef?.hasAttribute('valKind'))
      daiElement.setAttribute('valKind', srcValRef.getAttribute('valKind')!);
    if (srcValRef?.hasAttribute('valImport'))
      daiElement.setAttribute(
        'valImport',
        srcValRef.getAttribute('valImport')!
      );
    edits.push({
      parent: doiElement!,
      reference: null,
      node: daiElement,
    });
  }

  const valTextContent = controlBlockReference(controlBlock);
  let valElement = daiElement.querySelector(`Val`);
  if (!valElement) {
    valElement = subscriberIED.ownerDocument.createElementNS(
      SCL_NAMESPACE,
      'Val'
    );
    valElement.textContent = valTextContent;

    edits.push({
      parent: daiElement!,
      reference: null,
      node: valElement,
    });
  } else {
    edits.push({
      parent: valElement!,
      reference: null,
      node: new Text(valTextContent ?? 'Unknown Control Block'),
    });
  }

  return edits;
}
```

However, this doesn't provide fine grained control and perhaps we could improve this API.

As a user of this function I may want to sometimes instantiate against a new supervision node (even if a supervision is empty the user may not want it to be used, it may be "reserved" for another purpose).

I may also want to specify the supervision instance (?) that I want this to be instantiated against in some cases (is this sane? Perhaps there are many ways a supervision instance could be specified...).

### PR DESCRIPTION
**Pull request by JakobVogelsang**
Closes #20

### CODE REVIEW

**Comment by danyill** [comment]
Finally I am looking at this :tada:

**Review by danyill — CHANGES_REQUESTED** [review]
This is quite a large PR and I'm sorry I have not reviewed it in a timely manner :disappointed: . I appreciate your patience and this opportunity. :+1: :love_letter: 

This looks really good -- it is a surprisingly complex thing we are trying to do and it is very good to have it abstracted into `scl-lib`.

There seems to be just one issue - in that the supervision node is not always placed correctly - the rest are mostly questions and style comments. I have not carefully looked through the tests.

I have done a little test integration into oscd-supervision and it is looking good for the specific function we have here.

I did a little check on what we still need for supervisions to allow me to fully abstract out for oscd-supervision, oscd-subscriber-later-binding and oscd-subscriber-lb-*.

Here is the list I made:

* Add instantiation to `subscribe` function. Ideally when adding this, we would also have a `SubscribeOption` of `ignoreSupervision` (as we do for `unsubscribe`).
* https://github.com/OpenEnergyTools/scl-lib/issues/18
* Resolve https://github.com/OpenEnergyTools/scl-lib/issues/6
* Allow separate instantiation of a supervision element as per https://github.com/OpenEnergyTools/scl-lib/issues/20 (and fix the way we do deleting) and also provide an ability to check just "Can I instantiate a supervision LNode without necessary wanting to connect to a control block" which our current API does not allow.

I'm happy to do some of these but don't wish to interrupt your flow on this great work.

File: tLN/supervision/foundation.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  const serviceType = supervision.sourceControlBlock.tagName;
+  return serviceType === "GSEControl" ? "GoCBRef" : "SvCBRef";
+}
+
+export function supervisionLnClass(supervision: Supervision): "LGOS" | "LSVS" {
+  const serviceType = supervision.sourceControlBlock.tagName;
+  return serviceType === "GSEControl" ? "LGOS" : "LSVS";
+}
+
+/** Clojure returning unique inst attribute over for multiple subscriber IEDs */
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/foundation.ts:46]: Thank you will do :)

File: tLN/supervision/foundation.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    const lnClass = supervisionLnClass(supervision);
+
+    const formLn = ied.querySelector(`LN[lnClass="${lnClass}"]`)!;
+    const lDevice = formLn.parentElement!;
+
+    const id = `${ied.getAttribute("name")}`;
+    if (!lnInstGenerators.has(id))
+      lnInstGenerators.set(id, lnInstGenerator(lDevice, "LN"));
+
+    return lnInstGenerators.get(id)!(lnClass);
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/foundation.ts:66]: True thx

File: tLN/supervision/createSupervisionEdit.ts [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    inst,
+  });
+  const openScdTag = createElement(subscriberIed.ownerDocument, "Private", {
+    type: "OpenSCD.create",
+  });
+  ln.appendChild(openScdTag);
+
+  const reference =
+    parent.querySelector(`:scope > LN[lnClass="${lnClass}"]:last-child`)
+      ?.nextElementSibling || getReference(parent, "LN");
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/createSupervisionEdit.ts:137]: You are right I would still not want to get rid of the fallback and would instead propose to re-write as follows:

```ts
...
const lastSupervisionLn = parent.querySelector(`:scope > LN[lnClass="${lnClass}"]:last-child`)
const reference = lastSupervisionLn ? lastSupervisionLn.nextElementSibling : getReference(parent, "LN").
```

What do you say about this proposal @danyill ?

[the remark this answers is not in the document]
Comment (danyill) [reply tLN/supervision/createSupervisionEdit.ts:137]: Unless I err it still doesn't seem right:

So within an LDevice, we might have the following cases:

**Case 1:**

```
LN0
LN[lnClass="??1"]
LN[lnClass="LGOS"] <insert after this one>
LN[lnClass="??2"]
```

**Case 2:**

```
LN0
LN[lnClass="??1"]
LN[lnClass="LGOS"] <insert after this one>
AccessControl
```

**Case 3:**
```
LN0
LN[lnClass="??1"]
LN[lnClass="LGOS"] <insert after this one>
```

**Case 4:**
```
LN0 <insert after this one>
```

**Case 5:**
```
LN0 <insert after this one>
AccessControl
```

We want to add after the last `LGOS`, which may be the last element if there is no `AccessControl`.

- Case 1 and Case 2 are dealt with in the existing code.
- Case 3 must be handled by passing null (i.e. add to the end of the parent).
- Case 4 and 5 should be handled with `getReference`

Your proposal doesn't handle Case 3 I think (where we would insert between `LN0` and the first `LN`).

How about something like:

```suggestion
const lastSupervisionLn = parent.querySelector(`:scope > LN[lnClass="${lnClass}"]:last-child`)
let reference

if (lastSupervisionLn) {
  reference = lastSupervisionLn.nextElementSibling ? lastSupervisionLn.nextElementSibling : null;
} else  {
  reference = getReference(parent, "LN")
}
```

[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/createSupervisionEdit.ts:137]: The property `nextElementSibling` returns `null` when the element is the last element [nextElementsSibling](https://developer.mozilla.org/en-US/docs/Web/API/Element/nextElementSibling). As such, the line 139 can be reduced to `reference = lastSupervision.nextElementSibling` and then I think we are down to:
1. If there is a `lastSupervisionLn` pick the `nextElementSibling` (including the option there is no in which case we return null)
2. there is no such thing as `lastSupervisionLn` and we then let `getReference` figure out the correct next element sibling.

[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/createSupervisionEdit.ts:137]: And then I think the two-liner that I proposed should be enough. The name `lstSupervisionLn` is maybe a bit strange.

[the remark this answers is not in the document]
Comment (danyill) [reply tLN/supervision/createSupervisionEdit.ts:137]: You are quite correct, thanks for the discussion :+1: I think `lastSupervisionLn` is fine

File: tLN/supervision/createSupervisionEdit.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+          ) === null ||
+          ln.querySelector(
+            `:scope > DOI[name="${cbRefType}"] > DAI[name="setSrcRef"] > Val`
+          )?.textContent === ""
+        );
+      }) ?? null
+  );
+}
+
+function createSupervisionLogicalNode(
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/createSupervisionEdit.ts:109]: I would love to make it a separate PR. Would that be ok with you?

File: tLN/supervision/createSupervisionEdit.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  const cbRefType = type(supervision);
+
+  return (
+    Array.from(
+      supervision.subscriberIedOrLn.querySelectorAll(
+        `:scope > AccessPoint > Server LDevice > LN[lnClass="${lnClass}"]`
+      )
+    )
+      // filter already used available logical nodes
+      .filter((ln) => !userSupervisions?.has(ln))
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/createSupervisionEdit.ts:95]: The function can deal with multiple subscription supervisions at once. In these cases you need to save what supervisions have been set. The filter and the Set is filtering this.

File: tLN/supervision/canInstantiateSubscriptionSupervision.ts [hunk 6]
```diff
@@ -68,12 +41,14 @@ function exceedSupervisionLimits(supervision: Supervision): boolean {
...
+  if (existingSupervisionLogicalNode.length < parseInt(max, 10)) return true;
+
   const availableSupervisorSpots = existingSupervisionLogicalNode.filter(
     (ln) => !holdsValidObjRef(ln, type(supervision))
   );
 
   return (
-    existingSupervisionLogicalNode.length <= parseInt(max, 10) &&
+    existingSupervisionLogicalNode.length === parseInt(max, 10) &&
     availableSupervisorSpots.length > 0
```
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tLN/supervision/canInstantiateSubscriptionSupervision.ts:52]: Very true thank you!

**Comment by JakobVogelsang** [comment]
@danyill When this PR is in there is no running work and as such I would love if you would do some polishing.

**Comment by danyill** [comment]
Thank you for making these changes.

It looks good to me with the exception that we've lost a bit of code coverage by not testing the insertion order fallback:

image

**Review by danyill — APPROVED** [review]

**Pull request merged**